File: FStudyForum.Client/src/services/FeedService.ts
import { CreateFeed, Feed } from "@/types/feed";
import api from "./api";
import { ResponseWith, Response } from "@/types/response";

const getFeed = async (name: string) => {
  const response = await api.get<ResponseWith<Feed>>(`/feed/${name}`);
  return response.data.data;
};

const getFeeds = async (pageNumber: number, pageSize: number) => {
  const response = await api.get<ResponseWith<Feed[]>>(
    `/feed/all?pageNumber=${pageNumber}&pageSize=${pageSize}`
  );

  return response.data.data;
};

const createFeed = async (payload: CreateFeed) => {
  const response = await api.post<Response>(`/feed/create`, payload);
  return response.data.message;
};

const deleteFeed = async (name: string) => {
  const response = await api.delete<Response>(`/feed/${name}`);
  return response.data.message;
};

const addTopicToFeed = async (name: string, topicName: string) => {
  const response = await api.post<Response>(`/feed/add`, {
    feedName: name,
    topicName
  });
  return response.data.message;
};

const removeTopicFromFeed = async (name: string, topicName: string) => {
  const response = await api.post<Response>(`/feed/remove`, {
    feedName: name,
    topicName
  });
  return response.data.message;
};

const FeedService = {
  getFeed,
  getFeeds,
  addTopicToFeed,
  removeTopicFromFeed,
  createFeed,
  deleteFeed
};

export default FeedService;

File: FStudyForum.Infrastructure/Services/FeedService.cs
using AutoMapper;
using FStudyForum.Core.Interfaces.IRepositories;
using FStudyForum.Core.Interfaces.IServices;
using FStudyForum.Core.Models.DTOs;
using FStudyForum.Core.Models.DTOs.Feed;
using FStudyForum.Core.Models.DTOs.Topic;
using FStudyForum.Core.Models.Entities;
using Microsoft.AspNetCore.Identity;

namespace FStudyForum.Infrastructure.Services;

public class FeedService : IFeedService
{
    private readonly UserManager<ApplicationUser> _userManager;
    private readonly IFeedRepository _feedRepository;
    private readonly ITopicRepository _topicRepository;
    private readonly IMapper _mapper;
    public FeedService(UserManager<ApplicationUser> userManager,
        IFeedRepository feedRepository,
        IMapper mapper, ITopicRepository topicRepository)
    {
        _userManager = userManager;
        _feedRepository = feedRepository;
        _mapper = mapper;
        _topicRepository = topicRepository;

    }

    public async Task AddTopicToFeed(string username, AddFeedDTO addFeedDTO)
    {
        var feed = await _feedRepository.GetFeed(username, addFeedDTO.FeedName)
             ?? throw new Exception("Feed not found");
        var topic = await _topicRepository.GetByName(addFeedDTO.TopicName)
            ?? throw new Exception("Topic not found");
        if (feed.Topics.FirstOrDefault(t => t.Name == topic.Name) != null)
            throw new Exception("Topic is existed in feed");
        feed.Topics.Add(topic);
        await _feedRepository.Update(feed);
    }

    public async Task CreateFeed(string username, CreateFeedDTO createFeedDTO)
    {
        var feed = await _feedRepository.GetFeed(username, createFeedDTO.Name);
        if (feed != null) throw new Exception("Feed is existed");
        var creater = await _userManager.FindByNameAsync(username)
            ?? throw new Exception("User not found");
        await _feedRepository.CreateFeed(creater, createFeedDTO);
    }

    public async Task DeleteFeed(string username, string feedName)
    {
        var feed = await _feedRepository.GetFeed(username, feedName)
            ?? throw new Exception("Feed not found");
        await _feedRepository.Delete(feed);
    }

    public async Task<FeedDTO> GetFeed(string username, string feedName)
    {
        var feed = await _feedRepository.GetFeed(username, feedName)
            ?? throw new Exception("Feed not found");
        return new FeedDTO()
        {
            Id = feed.Id,
            Name = feed.Name,
            Author = username,
            Description = feed.Description,
            Topics = _mapper.Map<TopicDTO[]>(feed.Topics)
        };
    }

    public async Task<PaginatedData<FeedDTO>> GetFeeds(string username, QueryFeedDTO query)
    {
        var result = await _feedRepository.GetFeeds(username, query);
        var feedDTOs = new List<FeedDTO>();
        foreach (var feed in result.Data)
        {
            feedDTOs.Add(new FeedDTO()
            {
                Id = feed.Id,
                Name = feed.Name,
                Author = username,
                Description = feed.Description,
                Topics = _mapper.Map<TopicDTO[]>(feed.Topics)
            });
        }
        return new(feedDTOs, result.TotalCount);

    }

    public async Task RemoveTopicFromFeed(string username, RemoveFeedDTO removeFeedDTO)
    {
        var feed = await _feedRepository.GetFeed(username, removeFeedDTO.FeedName)
              ?? throw new Exception("Feed not found");
        var topic = feed.Topics.FirstOrDefault(t => t.Name == removeFeedDTO.TopicName)
            ?? throw new Exception("Topic isn't existed in feed");
        feed.Topics.Remove(topic);
        await _feedRepository.Update(feed);
    }
}
